### PR TITLE
Typing for React Router 0.13

### PR DESCRIPTION
**Pull request by duncanmak**
This is a new typing for React Router 0.13.

The older copy has been moved to the legacy folder.

Whenever the React typings get updated to include `context` type information, this typing needs to be updated to reflect that change. Currently, `Router.Context` can only be used via casting, an example can be found inside `react-router-test.ts`.

### CODE REVIEW

**Comment by vvakame** [comment]
@pspeter3 @vsiao could you review this PR?
@duncanmak please fix Travis CI failed. https://travis-ci.org/borisyankov/DefinitelyTyped/builds/59487899

File: react/react.d.ts [hunk 1]
```diff
@@ -3,7 +3,7 @@
 // Definitions by: Asana <https://asana.com>, AssureSign <http://www.assuresign.com>
 // Definitions: https://github.com/borisyankov/DefinitelyTyped
 
-declare module "react" {
+declare module React {
```
Comment (vsiao) [inline react/react.d.ts]: Please don't do this. `react-global.d.ts` is provided for this purpose.

Comment (duncanmak) [inline react/react.d.ts]: How do I reference `react-global.d.ts`?

**Comment by duncanmak** [comment]
I took out the commit that exported the `React` module and fixed up the `legacy` file paths.

File: react/react.d.ts (hunk 1 above)
Comment (duncanmak) [inline react/react.d.ts]: I fixed the PR.

What's the thinking behind having multiple `.d.ts` files?

Comment (vsiao) [inline react/react.d.ts]: 1) Declaring a global React module means that React may be referenced even when it hasn't been explicitly imported.
2) IntelliJ doesn't like it (I don't personally use IntelliJ with react/typescript, but see https://github.com/borisyankov/DefinitelyTyped/pull/3959 for more context)

**Comment by duncanmak** [comment]
I cleaned up the patches and we're down to one error:
`react-router/react-router-test.ts(71,10): error TS2339: Property 'contextTypes' does not exist on type 'typeof Messages'.`

I don't think this can be avoided, perhaps @vsiao has some ideas for what we can do here.

**Comment by vsiao** [comment]
Do you absolutely need to use context? It's not officially supported so I haven't fleshed out the types for it. Try setting `contextTypes` as a `static` property on `Messages`?

**Comment by duncanmak** [comment]
@vsiao Adding a static property breaks compilation - when I add that in, I get:

` 2345 Argument of type 'typeof Messages' is not assignable to parameter of type 'ComponentClass<any>'.`.

Here's a component definition in full:

```
class Messages extends React.Component<any, any> {
    context: { router: Router.Context };
    static contextTypes  = { router: React.PropTypes.func };
    DisplayName = 'Messages';
    render = () => DOM.p ({}, this.context.router.getCurrentParams().messageId);
}
```

This is most definitely related to https://github.com/borisyankov/DefinitelyTyped/issues/3641.

I haven't figured out how React Router use the `contextTypes` definition to inject the router yet - either way, without it, I get this error at runtime:

`test.ts:61 Uncaught TypeError: Cannot read property 'getCurrentParams' of undefined `

Note: The code in the PR right mirrors the approach shown in the React Router docs - https://github.com/rackt/react-router/blob/master/UPGRADE_GUIDE.md

**Comment by vsiao** [comment]
You'll need to type-annotate `contextTypes` with `React.ValidationMap<Context>` where `Context` is an interface that you define. The type errors given aren't very helpful at the moment, sorry about that. See https://github.com/borisyankov/DefinitelyTyped/blob/master/react/react-tests.ts#L77-79 for an example.

**Comment by duncanmak** [comment]
Thanks for working with me through this, @vsiao.

One difference now that I'm using `react-global.d.ts` - I don't have `import React = require('react')` at the top anymore, and with that, even though the TS -> JS compilation works, I see this error at runtime:

`Uncaught ReferenceError: React is not defined`.

**Comment by duncanmak** [comment]
Other than the error above, if I revert back to using an exported version of `react.d.ts`, the following component definition does work:

```
class Messages extends React.Component<any, any> {
    static contextTypes: React.ValidationMap<Router.Context> = {
        router: React.PropTypes.func
    }

    DisplayName = 'Messages';
    render = () => DOM.p ({}, this.context.router.getCurrentParams().messageId);
}
```

So, yay, progress!

**Comment by duncanmak** [comment]
Have we abandoned this pull request?

**Comment by vvakame** [comment]
@vsiao can i merge this PR?
@duncanmak please resolve merge conflict.

**Comment by vsiao** [comment]
I'd like @pspeter3 to take a look since he knows more about React Router :)

File: react-router/react-router.d.ts [hunk 2]
```diff
@@ -1,278 +1,171 @@
...
     }
-    
-    interface TransitionStaticLifecycle {
-        willTransitionTo?(
+
+    //
+    // Renderable Components
+    //
+    interface RouteHandler extends React.ComponentClass<any> {
+        willTransitionTo(
```
Comment (vsiao) [inline react-router/react-router.d.ts]: Is this right? I don't see `willTransitionTo` or `willTransitionFrom` in https://github.com/rackt/react-router/blob/master/modules/components/RouteHandler.js

File: react-router/react-router.d.ts [hunk 3]
```diff
@@ -1,278 +1,171 @@
...
+    interface LinkProp {
+        activeClassName?: string; // defaults to 'active'
+        to: string | Route;
+        params?: any;
+        query?: any;
+        activeStyle?: any;
+        onClick?: Function;
+    }
+
+    var Link: React.ComponentClass<LinkProp>;
```
Comment (vsiao) [inline react-router/react-router.d.ts]: Would prefer
```ts
class Link extends React.Component<LinkProps, {}> {
}
```
if that works

File: react-router/react-router.d.ts [hunk 4]
```diff
@@ -1,278 +1,171 @@
...
+        path?: string;
+        handler: Function;
+        ignoreScrollBehavior?: boolean
+    }
+
+    type Route = React.ReactElement<RouteProp>;
+
+    var Route:         React.ComponentClass<RouteProp>;
+    var DefaultRoute:  React.ComponentClass<RouteProp>;
+    var NotFoundRoute: React.ComponentClass<RouteProp>;
```
Comment (vsiao) [inline react-router/react-router.d.ts]: same for these component classes

**Comment by pspeter3** [comment]
I agree with @vsiao's comments in the code. I'm also interested why React.createElement needed to be overridden.

**Comment by vvakame** [comment]
I merged #4898 .

**Pull request closed without merging**